plugins/riscv/riscv_main.ml

```diff
@@ -0,0 +1,479 @@
+open Core_kernel.Std
+open Bap.Std
+open Or_error
+open Or_error.Monad_infix
+
+open Riscv_types
+open Riscv_utils
+
+module Insn = Disasm_expert.Basic.Insn
+module Branch = Riscv_branch
+module Env = Riscv_env
+
+module Riscv = struct
+  (** Defines the register map *)
+  module CPU = struct
+    let mem = Var.create "mem" @@ mem32_t `r8
+    let reg name = Var.create name reg32_t
+    let regs pref = Array.init ~f:(fun i -> reg @@ sprintf "%s%d" pref i)
+    let zero = reg "zero"
+    let r = regs "r" 31
+
+    let ra = reg "r1"
+    let sp = reg "r2"
+    let gp = reg "r3"
+    let tp = reg "r4"
+    let fp = reg "r8"
+
+    let pc = reg "pc"
+
+    let gprs = Array.concat [
+        r;
+        [|zero; gp; tp; sp; fp; ra|] ;
+    ]
+
+    let gpr = Array.to_list gprs |> Var.Set.of_list
+
+    let reg_of_name name =
+      let name = String.lowercase name in
+      Array.find gprs ~f:(fun reg ->
+      Var.name reg = name)
+
+    (* the problem is that RISC-V doesn't have flags at all,
+       but we can just pretend that they are. They will not
+       be used.
+       Seems that a drawback of some BAP architecture...
+       *)
+    let flag n = Var.create n bool_t
+    let zf = flag "zf"
+    let cf = flag "cf"
+    let vf = flag "vf"
+    let nf = flag "nf"
+    let never _ = false
+    let is_reg = Set.mem gpr
+    let is_flag = never
+    let is_zf = never
+    let is_cf = never
+    let is_vf = never
+    let is_nf = never
+    let is_sp v = Var.same v sp
+    let is_bp v = Var.same v fp
+    let is_mem v = Var.same v mem
+
+    let addr_of_pc m = Addr.(Memory.min_addr m ++ 4)
+  end
+
+  (** simplify an expression by applying constant folding *)
+  let simpl = Bil.fixpoint Bil.fold_consts
+
+  (** [reg op] is [Ok reg] if operand is a register with the same
+      name as reg *)
+  let reg = function
+    | Op.Imm _ | Op.Fmm _ -> Or_error.errorf "expected register"
+    | Op.Reg reg ->
+      let name = Reg.name reg in
+      match CPU.reg_of_name name with
+        | None -> invalid_argf "unknown register %s" name ()
+        | Some reg -> Ok reg
+
+  let imm = function
+    | Op.Reg _ | Op.Fmm _ -> Or_error.errorf "expected immediate"
+    | Op.Imm imm -> Ok imm
+
+  (** [rrr_type f d s t] uses lifter [f] for an rrr-type instruction with
+      arguments [d], [s], and [t]. *)
+  let rrr_type f d s t = match reg d, reg s, reg t with
```

Comment:
maybe it is better to apply `simpl` once at the toplevel lift function just before the the value is returned? In fact, you may simply omit the `simpl` function, as I will still apply simplification/normalization on later phases. Here it is mostly useful for debugging. Probably, we should add a `--simpl/--no-simpl` and other related options to `bap-mc`.